daemons/dss-sync/app.py

```diff
@@ -24,7 +33,97 @@ def process_new_syncable_object(event, context):
+                    if gs_bucket.get_blob(dest_blob_name) is None:
+                        compose_gs_blobs(gs_bucket, source_blob_names, dest_blob_name)
+                    compose_stage2_blob_names.append(dest_blob_name)
+            else:
+                parts_to_compose = range(1, msg["total_parts"] + 1)
+                compose_stage2_blob_names = ["{}.part{}".format(msg["dest_key"], p) for p in parts_to_compose]
+            context.log("Composing, stage 2")
+            compose_gs_blobs(gs_bucket, compose_stage2_blob_names, msg["dest_key"])
+            break
+        except AssertionError:
+            pass
+        time.sleep(5)
+
+@app.sns_topic_subscriber("dss-s3-mpu-ready")
+def complete_multipart_upload(event, context):
+    msg = json.loads(event["Records"][0]["Sns"]["Message"])
+    mpu = resources.s3.Bucket(msg["dest_bucket"]).Object(msg["dest_key"]).MultipartUpload(msg["mpu"])
+    while True:
+        context.log("Examining parts")
+        parts = list(mpu.parts.all())
+        if len(parts) == msg["total_parts"]:
+            context.log("Closing MPU")
+            mpu_parts = [dict(PartNumber=part.part_number, ETag=part.e_tag) for part in parts]
+            mpu.complete(MultipartUpload={'Parts': mpu_parts})
+            context.log("Closed MPU")
+            break
+        time.sleep(5)
+
+log_msg = "Copying {source_key}:{part} from {source_platform}://{source_bucket} to {dest_platform}://{dest_bucket}"
+platform_to_replica = dict(s3="aws", gs="gcp")
+@app.sns_topic_subscriber("dss-copy-parts")
+def copy_parts(event, context):
+    topic_arn = event["Records"][0]["Sns"]["TopicArn"]
+    msg = json.loads(event["Records"][0]["Sns"]["Message"])
+    blobstore_handle = dss.Config.get_cloud_specific_handles(platform_to_replica[msg["source_platform"]])[0]
+    source_url = blobstore_handle.generate_presigned_GET_url(bucket=msg["source_bucket"], object_name=msg["source_key"])
+    futures = []
+    gs = dss.Config.get_cloud_specific_handles("gcp")[0].gcp_client
+    with ThreadPoolExecutor(max_workers=4) as executor:
+        for part in msg["parts"]:
+            context.log(log_msg.format(part=part, **msg))
+            if msg["dest_platform"] == "s3":
+                upload_url = "{host}/{bucket}/{key}?partNumber={part_num}&uploadId={mpu_id}".format(
+                    host=clients.s3.meta.endpoint_url,
+                    bucket=msg["dest_bucket"],
+                    key=msg["dest_key"],
+                    part_num=part["id"],
+                    mpu_id=msg["mpu"]
+                )
+            elif msg["dest_platform"] == "gs":
+                assert len(msg["parts"]) == 1
+                dest_blob_name = "{}.part{}".format(msg["dest_key"], part["id"])
+                dest_blob = gs.get_bucket(msg["dest_bucket"]).blob(dest_blob_name)
+                upload_url = dest_blob.create_resumable_upload_session(size=part["end"] - part["start"] + 1)
```

Comment:
Does the client (our code, in this case) even know how to deal with resuming an upload?